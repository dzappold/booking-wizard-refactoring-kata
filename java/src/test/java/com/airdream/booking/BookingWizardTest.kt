package com.airdream.booking

import com.github.stefanbirkner.systemlambda.SystemLambda.tapSystemErrAndOutNormalized
import com.github.stefanbirkner.systemlambda.SystemLambda.withTextFromSystemIn
import org.approvaltests.Approvals.verify
import org.junit.jupiter.api.Test

class BookingWizardTest {

    @Test
    fun `outward flight with 2 passengers`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "1",
                "Frankfurt",
                "London",
                "10/12/2024",
                "2",
                "Michael Collins",
                "Sue Morrison",
                "Yes"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }

    @Test
    fun `round trip flight with 4 passengers`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26/09/2024",
                "4",
                "James Hetfield",
                "Marie",
                "Michael Collins",
                "Sue Morrison",
                "Yes"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }

    @Test
    fun `round trip flight with 4 passengers no confirmation`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26/09/2024",
                "4",
                "James Hetfield",
                "Marie",
                "Michael Collins",
                "Sue Morrison",
                "No"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }

    @Test
    fun `invalid trip`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "7",
                "abc",
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26/09/2024",
                "4",
                "James Hetfield",
                "Marie",
                "Michael Collins",
                "Sue Morrison",
                "No"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }

    @Test
    fun `departure in the past`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "7",
                "abc",
                "2",
                "Frankfurt",
                "London",
                "26/07/1999",
                "26/07/2024",
                "26/09/2024",
                "4",
                "James Hetfield",
                "Marie",
                "Michael Collins",
                "Sue Morrison",
                "No"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }

    @Test
    fun `invalid departure date`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "7",
                "abc",
                "2",
                "Frankfurt",
                "London",
                "26#07--1999",
                "26/07/2024",
                "26/09/2024",
                "4",
                "James Hetfield",
                "Marie",
                "Michael Collins",
                "Sue Morrison",
                "No"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }
    @Test
    fun `return date in the past`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26/07/1999",
                "26/09/2024",
                "1",
                "Sue Morrison",
                "Yes"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }

    @Test
    fun `return date before departure`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26/07/2023",
                "26/09/2024",
                "1",
                "Sue Morrison",
                "Yes"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }
    @Test
    fun `return date invalid format`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26#07--2023",
                "26/09/2024",
                "1",
                "Sue Morrison",
                "Yes"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }
    @Test
    fun `invalid passenger number`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26/09/2024",
                "abc",
                "1",
                "Sue Morrison",
                "Yes"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }
    @Test
    fun `invalid confirmation`() {
        val output = tapSystemErrAndOutNormalized {
            withTextFromSystemIn(
                "2",
                "Frankfurt",
                "London",
                "26/07/2024",
                "26/09/2024",
                "1",
                "Sue Morrison",
                "something",
                "No"
            ).execute {
                val wizard = BookingWizard()
                wizard.start()
            }
        }
        verify(output)
    }
}
