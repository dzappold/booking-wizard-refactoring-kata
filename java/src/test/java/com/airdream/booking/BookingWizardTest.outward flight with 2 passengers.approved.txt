Hello, welcome to Airdream, the airline of your dreams
What kind of flight do you want to book?
[1] outward
[2] round trip
Select your departure city
Select your arrival city
Select your departure date (DD/MM/YYYY)
Please select the number of passengers
Please enter the passenger number 1 first name and last name
Please enter the passenger number 2 first name and last name
Your booking summary
Flight 1: Departure at Tue Dec 10 00:00:00 CET 2024 from Frankfurt and arrival in London
Passengers:
Michael Collins
Sue Morrison
Do you confirm? [Yes/No]
Thank you for your booking, we wish you an excellent trip to London
